### PR TITLE
Fix threading.Condition with monkey-patching on Python 3.3 and newer

### PR DESCRIPTION
**Pull request by haypo**
It fixes monkey-patching threading.Condition on Python 3.3 and newer:
fix issue #185.

### CODE REVIEW

**Comment by haypo** [comment]
Is it ok to change Semaphore.acquire() behaviour for negative timeout? Any issue with the backward compatibility? The idea is to have a behaviour to the Python stdlib.

**Comment by haypo** [comment]
Ping? Is someone available to review this change? Is it ok?

**Comment by jstasiak** [comment]
@haypo I'll review this tomorrow, sorry for the wait

**Comment by jstasiak** [comment]
@haypo The negative timeout value seems to be working already, do you have a test case showing a failure?

**Comment by haypo** [comment]
> @haypo The negative timeout value seems to be working already, do you have a test case showing a failure?

My first commit changes the behaviour of eventlet.semaphore.Semaphore.acquire() when you pass a negative timeout. Before, Semaphore.acquire(timeout=-1) was non-blocking. After, Semaphore.acquire(timeout=-1) becomes blocking, same behaviour than Python 3 lock.

In Python 2, threading.Lock.acquire() has no timeout parameter. So RLock only calls acquire() (on its internal lock) with one parameter: blocking.

In Python 3, threading.Lock.acquire() got a new timeout parameter: def acquire(blocking=True, timeout=None). RLock calls aquire() (on its internal lock) with two parameters: blocking and timeout.

In Python 3, threading.Lock.acquire(timeout=-1) blocks, whereas eventlet.semaphore.Semaphore.acquire(timeout=-1) is currently non-blocking.

eventlet monkey-patching is not supposed to modify the behaviour of patched functions. I wasn't the case in Python 2, since Lock.acquire() had no timeout parameter. It's now the case, since Python 3 also uses the second timeout parameter.

In short, my changes fix two different Python 3 bugs:
- fix monkey-patching on threading.Lock to have the same behaviour on Lock.acquire(timeout=-1)
- fix RLock, it didn't work with monkey-patching on Python 3.3+

**Comment by haypo** [comment]
FYI I failed to find a call to Semaphore.acquire() which uses the timeout parameter in eventlet source code. In eventlet, the timeout doesn't look to be used.

The timeout parameter is documented:
http://eventlet.net/doc/modules/semaphore.html#eventlet.semaphore.Semaphore.acquire

The behaviour on negative timeout is not documented.

**Comment by haypo** [comment]
Ping?

**Comment by temoto** [comment]
@haypo I've rebased it on recent master, now `tests.patcher_test:Threading.test_threading` fails with timeout. Could you look at it? https://github.com/eventlet/eventlet/tree/semaphore_timeout

**Comment by haypo** [comment]
> @haypo I've rebased it on recent master,

Cool!

> now tests.patcher_test:Threading.test_threading fails with timeout. Could you look at it? https://github.com/eventlet/eventlet/tree/semaphore_timeout

It's not a regression of recent changes in eventlet. I'm not sure, but I guess that I used Python 3.3 to work on my patch, whereas Python 3.4 requires extra changes. Python 3.4 adds a new threading.Thread._tstate_lock attribute which doesn't work with monkey-patching, because it relies on private C code which releases the new lock at thread exit (when its Python thread state is destroyed).

Well, see the pull request https://github.com/eventlet/eventlet/pull/211 for the details, I proposed a fix.

**Comment by haypo** [comment]
The changeset "Disable the thread state lock" fixes the threading support on Python 3.4 when eventlet is used with monkey-patching.

Python 3.4 introduces threading.Thread._tstate_lock to fix a race condition, see:
- https://hg.python.org/cpython/rev/d52b68edbca6
- http://bugs.python.org/issue18808

I'm not really proud of this change: patching a Thread instance after its creation is not the most elegant fix.

I tried to create a Thread subclass in eventlet.green.threading, but I got a lot of new issues. If the subclass is declared in eventlet.green.threading, the parent Thread class uses functions of the original threading module, instead of using patched functions by eventlet. It's unclear to me how symbols are added and patched in modules. I noticed that the threading module of the standard library is imported multiple times. The implementation of monkey-patching is really complex :-/

Currently, the best solution is to patch Thread instance in the patched thread.start_new_thread() function. The problem is to retrieve the instance. My heuristic checks if the wrapped function is the bounded Thread.run() method, and I retrieve the instance from function.__self__.

I'm not yet sure that I really understood the problem. At least, "tox -e py34-selects" pass.

**Comment by haypo** [comment]
I rewrite the first commit: in fact, we don't need to support negative timeout, we only need to accept timeout=-1 for Semaphore.accept(). If the timeout value is None or -1, the timeout is ignored, only the blocking parameter is used:
- acquire(timeout=None)/acquire(timeout=-1) is the same than acquire(blocking=True): blocking
- acquire(blocking=False, timeout=None)/acquire(blocking=False, timeout=-1) is the same than acquire(blocking=False): non-blocking

File: eventlet/semaphore.py [hunk 1]
```diff
@@ -78,9 +78,18 @@ def acquire(self, blocking=True, timeout=None):
...
         same thing as when called without arguments, and return true.
+
+        Timeout value must be strictly positive.
         """
-        if not blocking and timeout is not None:
-            raise ValueError("can't specify timeout for non-blocking acquire")
+        if timeout == -1:
+            timeout = None
+        if timeout is not None and timeout < 0:
+            raise ValueError("timeout value must be strictly positive")
```
Comment (temoto) [inline eventlet/semaphore.py]: Exception says "strictly positive" ( x > 0 ), while code allows timeout=0 just fine.

Comment (haypo) [inline eventlet/semaphore.py]: > Exception says "strictly positive" ( x > 0 ), while code allows timeout=0 just fine.

Hum... yes... I copied code from the C implementation of threading.Lock, to try to be as close as possible to threading.Lock.

Example:

```
$ python3
Python 3.4.1 (default, Nov  3 2014, 14:38:10) 
>>> import threading
>>> l=threading.Lock()
>>> l.acquire(timeout=-2)
...
ValueError: timeout value must be strictly positive
>>> l.acquire(timeout=0)
True
>>> l.acquire(timeout=0)
False
```

**Comment by haypo** [comment]
temodo wrote "Exception says "strictly positive" ( x > 0 ), while code allows timeout=0 just fine." but I copied the error message from Python.

Do you expect that I change my patch, or are you ok to mimick Python lock?

**Comment by temoto** [comment]
Quite surprised that Python code has this code/comment mismatch. Yeah, leave it same as stdlib.

**Comment by haypo** [comment]
> Quite surprised that Python code has this code/comment mismatch. Yeah, leave it same as stdlib.

Oh, so did you review my patch? Does it look ok? Should I change something?

**Comment by temoto** [comment]
LGTM, running last round of tests now.

**Comment by temoto** [comment]
I took the liberty of sqashing PEP-8 fixes into your commits.

**Comment by haypo** [comment]
> I took the liberty of sqashing PEP-8 fixes into your commits.

No problem. I only care of having eventlet running on python 3.3 ;-)

**Comment by temoto** [comment]
It's merged into master 390e71d89883649dd55fbb70dc8686bf99e2a2e2 thank you!
Please test if it works and drop a line.

**Comment by jstasiak** [comment]
@hardys @temoto Can this issue be closed now?

**Comment by haypo** [comment]
> @hardys @temoto Can this issue be closed now?

A change fixing this issue was merged, so yes, the issue can be closed.

**Comment by jstasiak** [comment]
Thanks @haypo!
